File: javascript/bowling/180606/bowling_180606.js
export default class Bowling {

    constructor() {
        this.rolls = []
    }

    roll(pins) {
        this.rolls.push(pins);
    }

    isStrike(index) {
        return (this.rolls[index] == 10);
    }

    isSpare(index) {
        return (this.rolls[index] + this.rolls[index + 2]) == 10;
    }

    strikeBonus(index) {
        return 10 + this.rolls[index + 1] + this.rolls[index + 2];
    }

    spareBonus(index) {
        return 10 + this.rolls[index + 1];
    }

    score() {
        let totalScore = 0;
        let index = 0;
        for (let frame = 0; frame < 10; frame ++) {
            if (this.isStrike(index)) {
                totalScore += this.strikeBonus(index);
                index ++;
            } else if (this.isSpare(index)) {
                totalScore += this.spareBonus(index);
                index += 2;
            } else {
                totalScore += this.rolls[index] + this.rolls[index + 1];
                index += 2;
            }
        }
        return totalScore;
    }

}

File: bowling/practice/180606/bowling_180606.py
class BowlingGame():

    def __init__(self):
        self.rolls = []

    def roll(self, pins):
        self.rolls.append(pins)

    def isSpare(self, roll):
        return 10 == (self.rolls[roll] + self.rolls[roll + 1])

    def isStrike(self, roll):
        return 10 == self.rolls[roll]

    def spareBonus(self, roll):
        return 10 + self.rolls[roll + 2]

    def strikeBonus(self, roll):
        return 10 + self.rolls[roll + 1] + self.rolls[roll + 2]

    def score(self):
        totalScore = 0
        roll = 0
        for frame in range(10):
            if self.isStrike(roll):
                totalScore += self.strikeBonus(roll)
                roll += 1
            elif self.isSpare(roll):
                totalScore += self.spareBonus(roll)
                roll += 2
            else:
                totalScore += (self.rolls[roll] + self.rolls[roll + 1])
                roll += 2
        return totalScore